src/sign.ml

```diff
@@ -264,18 +269,36 @@ let read : string -> t = fun fname ->
 (* NOTE here, we rely on the fact that a marshaled closure can only be read by
    processes running the same binary as the one that produced it. *)
 
-(** [add_rule sign sym r] adds the new rule [r] to the symbol [sym].  When the
-    rule does not correspond to a symbol of signature [sign],  it is stored in
-    its dependencies. *)
-let add_rule : t -> sym -> rule -> unit = fun sign sym r ->
-  sym.sym_rules := !(sym.sym_rules) @ [r];
-  if sym.sym_path <> sign.sign_path then
-    let m =
-      try PathMap.find sym.sym_path !(sign.sign_deps)
-      with Not_found -> assert false
-    in
-    let m = (sym.sym_name, r) :: m in
-    sign.sign_deps := PathMap.add sym.sym_path m !(sign.sign_deps)
+(** [add_rules s r] adds a bunch of rules [r] to its attached symbol to the
```

Comment:
Is the meaning of your comment @rlepigre is to remove the decision tree computed from the object file (.lpo or .dko)? If it is the case, what you are asking, is that every time a defined symbol is imported in the signature, its decision tree is computed, again? If so, this won't scale for bigger libraries or if encodings to Dedukti becomes more complex. With Universo, I was able to see the limits of decision trees of Dedukti and this is why we introduced laziness to evaluate decision trees. Also, Inductive types and subtyping in Coq probably needs symbols declared in the logic file which are refined every time an inductive type is defined. Recompute such decision tree every time you import a file which does not declare an inductive type would be very anyone in that case since the number of rules for a same symbol might be big.

Reply:
@rlepigre, do you mean to recompute decision trees or to extend Sign.link to decision trees?

In any case, @francoisthire is right: enforcing physical equality of symbols at loading may not scale. We had some discussion about this a few days ago with @barras . It would be good to rediscuss the advantages and inconvenients of this choice. If I remember well, the main reason is that the Terms.sym data structure contains the rules. An alternative would be to have a map/hashtbl from symbol names to rules. In any case, it will be good @rlepigre to add some comment in sign.ml explaining this choice.

By the way I remark that `Sign.sign_deps : (string * rule) list PathMap.t ref` instead of `Sign.sign_deps : (string * rules) list PathMap.t ref` as several rules can be given at the same time (rule ... and ...). Shouldn't this be changed?

Reply:
What I am here proposing is an easy fix so that the PR does not break Lambdapi. Of course, it will probably be useful to not recompute decision trees every time we load a file (note that with what I suggest, the generation is only done once for every symbol, not once every rule or group of rules, so it is not too bad and it might not even be worth doing better than that). I think it is a good idea to move on quickly with this PR, and not debate such details at the moment (though we should keep what you said in mind @francoisthire). Fixing that will be easy in a second step. Same thing for the representation of symbols @fblanqui, I think it might be worth reconsidering this choice, but this PR is not the right place.

@fblanqui for the `Sign.sign_deps` field, I think it should remain as it was before. No need for nested lists here. This does not prevent us to do the minimal work when recomputing decision trees for the extra symbols (of course, rules on the same symbol can be defined in two different groups, so what you propose does not improve anything here).

Reply:
And @fblanqui, I forgot, enforcing physical equality at loading time is not that expensive. It only involves a (almost) linear traversal of terms, which is probably not that much more expansive than unmarshaling.